### ISSUE TITLE
once operator validity

### ISSUE THREAD

**Issue by gammaw** [savonet/liquidsoap#1544]
I wonder what the "validity" of the `once` operator is (see related issue:  https://github.com/savonet/liquidsoap/issues/1452).  

We have a radio setup with the following statement in the heart of our configuration. The focus is on `once`. We're using it successfully to play tracks that recur weekly. Every playlist contains exactly one track, which must play exactly once. It works like this every week, over and over (without restarting/reloading). My question is why that happens? Why will the playlist be available again after a week? Would it be available the next day, for example? 

```
radio = switch(track_sensitive=false, 
    transitions=[
        transition_into_show(playlist_jingle_station_id),
        (...),], 
    transition_length=60., #max duration of jingle (60s)
    [
        ({!live_enabled}, live), #Live DJ          
        ({1w and 10h11m-11h11m}, once(playlist_show1)),# Monday
        (...),
        ({true}, radio) 
    ])
```

In the other issue I mentioned, `playlist.reloadable` was proposed. I wouldn't know where and when I could reload the playlist. We would need to reload it right after it stopped playing, so that it becomes available again. The transition would be a good place, but there I don't have a handle of the playlist. Am I missing something? 

Any help is appreciated!

**Comment by smimram** [comment]
I don't think your script is supposed to work as is: `once(s)` is supposed to play one track of `s` and never reload (be it after 1 hour or 1 week). As for the reason why this is not the case, I would have to check. I'd say that the proper way to achieve what you want would be to replace
```
({1w and 10h11m-11h11m}, once(playlist_show1)),
```
by
```
(predicate.at_most(1, {1w and 10h11m-11h11m}), playlist_show1),
```
if you are using Liquidsoap 2.0.

**Comment by gammaw** [comment]
Thanks @smimram! 
To be honest, we're also confused why our configuration works at all. We're using Liquidsoap 1.4.2, it it helps. And [this](https://github.com/mmmnmnm/lahmacun/blob/master/azuracast/liquidsoap_custom.liq) is our full config. 

For you proposal: I guess I get the point, but I see some open issues: 
- How do we make sure that the single track inside the playlist (this is by design the case for us) will not restart the track when it finishes within the time frame (10h11m - 11h11m)? This is the behaviour we'd like to have. 
- How will the playlist/predicate be reloaded so that it becomes available again? Typically, we need to reload a playlist after a week, but sometimes every day, or even less. 

Btw, is Liquidsoap 2.0 already available? I found `predicate.at_most` in your recent commit: https://github.com/savonet/liquidsoap/commit/2d2127c46384bbdb31889e7cf7117b881b518d47

**Comment by smimram** [comment]
Let met try to explain more. `predicate.at_most(1, p)` means "when you are asked the value, you should be true the first time where `p` is true" and this is reset each time `p` becomes false. So `predicate.at_most(1, {10h - 11h})` will be true the first time it is asked in the timespan 10h - 11h (and then it will be false until it is asked after 11h and then it is asked again within 10h-11h).

Does this make sense? This should answer your second point (I am not sure I understand the first point)

Liq 2.0 is not yet available (apart from the git). Hopefully it will be soon. Until then, copying the code you pointed in your script should be enough.

**Comment by gammaw** [comment]
Thanks for the clarification. I'm starting to get it, I think. For my second point, and to use an example to doublecheck: 

Say `playlist_show1` contains a single audio file (e.g., `track.mp3` which is 20-minute long). With `(predicate.at_most(1, {1w and 10h-11h}), playlist_show1)`: 
- `track.mp3` will start playing at 10h (on Monday),
- `track.mp3` will play until the end (until 10:20), 
- `track.mp3` will not play again, because this would require the predicate to be true again (?),
- `track.mp3` will play again after a week, again at 10h. 

Is this correct?

And what would happen if `track.mp3` is longer than 60 minutes and we have `track_sensitive = false`. Would that mean that the track will be interrupted (faded out) at 11h? Because this is the behaviour what we'd like to have (and what we have now). 

And what if `playlist_show1` has two tracks (say `track1.mp3` and `track2.mp3`, 20 minutes each)? Would both tracks play play within the 1 hour timeframe? 

The basic difference, as I understand, between our current approach (`once(playlist)`) and your proposal (`predicate.at_most`) is that the former restricts the availability of the playlist while the latter controls the validity of the "trigger" (via the predicate).

**Comment by smimram** [comment]
Yes you are correct!

BTW, I added `predicate.once` as an alias for `predicate.at_most(1, ...)`.

This strategy will be a complete failure if we have `track_sensitive=false` because the predicate will be queried at each frame (= buffer of 0.04 seconds) which means that you will actually only hear the first frame of your song. I am not really sure about the idea of playing one track in a non-track-sensitive world. If you really need this, I can think harder and try to come up with a decent solution.

If there are two tracks in `playlist_show1` (and you are track sensitive) only the first will be played: the predicate waits for `{1w and 10h-11h}` to be `false` before becoming `true` again, which means that it will say "ok" for the first track and then for the second track it will say "I am saying no, and will think again once we are after 11h".

You are correct with the source (here the playlist) vs trigger approach. The thing is that we cannot easily achieve what we want with the source only, because we need to have some information about when to reset the "once" and this information can only be extracted from the trigger...

**Comment by gammaw** [comment]
@smimram thanks for the explanation!

> This strategy will be a complete failure if we have track_sensitive=false because the predicate will be queried at each frame (= buffer of 0.04 seconds) which means that you will actually only hear the first frame of your song. I am not really sure about the idea of playing one track in a non-track-sensitive world. If you really need this, I can think harder and try to come up with a decent solution.

Indeed, `track_sensitive=false` is one of our main assumptions. The reason is that we use it to force overly long shows to end on time. In my above example `track.mp3` is in fact a one-hour show, but since it's produced externally, it may be shorter or longer than 60 minutes. That's why we depend on Liquidsoap 1) not to play the show again if it finishes before the time slot ends (hence the rationale for the `once` operator) and 2) to end the track if it exceeds the scheduled time slot (and we use a fade out transition in this case). Does is make sense? 

Now I see that `once` is not meant for that but it has served us well (it still does), but now we would need more control over it to make the playlist available again (e.g., every hour or every other day). Also, we don't feel safe not knowing why our solution works at all :)

**Comment by smimram** [comment]
Maybe a solution could be to use the `delay` operator: it ensures that we won't play two tracks of the same source before _n_ seconds. This looks like a good approximation of a "reloadable `once`", doesn't it?

**Comment by gammaw** [comment]
I love it! I see that `delay` operates on a source, so we could use `delay(3600, playlist_show1)` instead of `once(playlist_show1)` (in case of a 1-hour show) and that's it. The only limitation would be that the show cannot be scheduled within an hour after the track finished. The value must be conservatively set so hight (3600) if we don't want to make assumptions about the length of the track. Is this what you mean?

**Comment by smimram** [comment]
Yes, this was precisely my suggestion!

**Comment by gammaw** [comment]
@smimram I tried it and it principally works, except one important detail: if the audio (the single audio file in the playlist) is "too long" for the slot, the audio will continue there where it ended previously when the playlist is scheduled next. 

In the following example, say the audio is 6 minutes long, it will end after 5 minutes at 13h20m, and when the playlist is scheduled to restart at 13h35m, the audio will continue at 5 minutes, play to the end, and the playlist will end at 13h36m. 

`({1w and (13h15m-13h20m or 13h35m-13h45m)}, delay(540.0,playlist_mc5)),`

Instead, we'd like that the playlist "starts over" every time the playlist is scheduled to play again. Do you have an idea how to achieve it? 

I tried using `skip` in the transition but it didn't work (if I use it before, then it will skip the track and end without playing it, if I use it after, then I think fading wouldn't work). These is the transition function we're using (in both directions, for entering and leaving the playlist): 

```
def transition(j,a,b)
  sequence(merge=true, [fade.final(duration=5.,a),j,b])
end
```

**Comment by gammaw** [comment]
An addition to my previous comment: I thought I had a good idea when I found the `fade.skip` operator (we're running v1.4.2), which seems to be doing something different but the code helped to come up with a custom "fade & skip" function (see below). Result: using the same example from above, the skip happens (I can see it in the log) but the audio won't play at all for the second time (at 13h35m). I tried with double skipping (hoping that the playlist would start over) but it didn't work either. What do you think? 

Note: interestingly, when I schedule the playlist for a third time, it would play the file again. 

```
def fade.final_skip(~id="fade.final",~duration=3.,~type="lin",s) =
  def skip()
    log("Fade finished executing. Calling skip now") 
    source.skip(s);
  end

  fn = mkfade(start=1.,stop=0.,type=type,duration=duration,on_done=skip,s)
    should_play = ref true

  def fn() =
    v = fn()
    if v == 0. then
      should_play := false
    end
    v
  end

  s = amplify(id=id,fn,s)
  switch(track_sensitive=false,[({!should_play},s)])
end
```

**Comment by smimram** [comment]
Thinking of your situation I wrote #1560, but it will be for 2.0 at best...

With an "old" Liquidsoap, maybe can you run a thread with `add_timeout` to regularly check (say every minute) if the condition has become `false` and in this case skip the source? I know it's not pretty but it should work...

**Comment by gammaw** [comment]
Thanks @smimram for your effort on 2.0! I'm not sure about the "watchdog" workaround (threads with `add_timeout`) because we have 50+ playlists. Also, I'm not sure that the problem is with `skip`. 

I was wondering why my solution with the custom `fade.final_skip` script doesn't work. The `skip` seems to execute after the fade-out. What I don't understand why the playlist will never play again (after the first time of playing). As far as I know, a playlist will start over after it played the last track. 

> Note: interestingly, when I schedule the playlist for the third time, it would play the file again.

I guess I was wrong here because the file will never play again. At least it was the case when I tried today. 

Let me show you my example again: 

`({1w and (13h15m-13h20m or 13h35m-13h45m or 13h55m-14h05m)}, delay(540.0,playlist_mc5))`

`playlist_mc5` contains a single file (`track.mp3`) and it is 6-minute long. What happens is that the file starts playing at 13h15m, `fade.final_skip` executes at 13h20m (it fades out the audio and skips the source), and the file never plays again (I would expect it to play again at 13h35m and 13h55m). What happens instead is that the playlist is "entered" (I can see the transitions, jingles etc.) but `track.mp3` is not played. A related entry I see in the log is this: 

`Switch to delay_8193 with transition.`

Long story short, do you know why `track.mp3` is played only once and not three times?

**Comment by gammaw** [comment]
@smimram do you have any hint for that? I would happily experiment further but I'm kind of stuck at the moment. Because of what I described in my previous comment, I don't know if it's an issue with `skip` or it's wrong playlist handling on our side. Any help is appreciated.

**Comment by smimram** [comment]
I don't see why it should not be played 3 times :( Are you sure that some other condition is not `true` somewhere before in your switch which would prevent selecting this one?

**Comment by gammaw** [comment]
Thanks for the confirmation! I ran some more experiments and it turns out that when the track is not "too long", then it plays again and again as expected. Using a simplified example with just two rounds and assuming a single 6-minute long track in the playlist, I had the following results:

`({1w and (13h15m-13h20m or 13h35m-13h45m)}, delay(540.0,playlist_mc5)) #plays only at 13h15m`

but

`({1w and (13h15m-13h25m or 13h35m-13h45m)}, delay(540.0,playlist_mc5)) #plays at 13h15m and at 13h35m`.

The fade script only runs in the first case, the `skip` is called, and, for some reason, the track won't play on the second occasion. For me, it means, that the script changes something about the availability of the playlist. As I write before, I see "Switch to delay_8193 with transition." in the log at 13h35m, but the track is not played. 

Do you have an idea why this happens?

**Comment by gammaw** [comment]
I have a theory :) It seems to me that, if there is a fade, the delay time will start counting in the next "iteration". That is, if the track ends in the slot, then the delay will be triggered immediately, but, if the track needs to be faded out (and skipped), then the delay will be triggered the next time when the playlist is scheduled. Strangely, the `source.is_ready` predicate returns true but the track will not start until the delay time is passed. 

In the above example:

`({1w and (13h15m-13h20m or 13h35m-13h45m)}, delay(540.0,playlist_mc5)) #plays only at 13h15m`, 

the 6-minute track will be faded and skipped at 13h20m, but the delay will start counting at 13h35m (with `source.is_ready` being true). If we use a shorter delay, e.g., 5 minutes (300.0), then the track will start again at 13h40m. 

If this is correct, then `delay` turns out to be tricky for our use case. The dilemma: if we have a shorter-than-the-slot track, then we need a delay (to prevent the track from replaying), if we have a longer-than-the-slot track, then we don't need a delay (and it causes trouble later). 

What do you think? Don't you think that the `delay` operator should "detect" the `skip`? What do you think could be a way out?

**Comment by gammaw** [comment]
@smimram it's been some time and we've been busy with other stuff too. Now we'd like to revisit this and hopefully find a solution. Did you have a chance to check the last comment? Do you have a hint for us? Any help is appreciated.

**Comment by smimram** [comment]
You were right, `delay` did not correctly handle `skip`. This should now be corrected in 21b66f3d1fcf40e0df86bd04939e7f6b43c34876. Any chance that you could try the latest git version?

**Comment by gammaw** [comment]
Wow! I very much look forward to trying the fix! It's a little bit tricky as we run Liquidsoap inside an Azuracast instance but we should be able to plug in a fresh build there. I'll let you know once we have a result.

**Comment by gammaw** [comment]
@smimram you may be able to help, I'm a bit stuck. We finally managed to build Liquidsoap pinned to the commit you posted and we successfully mounted the file into our Azuracast instance. We did the build using OPAM on Ubuntu using the default dependencies as instructed in your installation manual. 

Running the new version we got a rather unexpected error. This one: 

```
At line 16, char 21-30:
Error 5: this value has type
  unit
but it should be a subtype of
  (id : _, mode : _, reload_mode : _, conservative : _, default_duration : _,
   length : _, _) -> _
```

The error message complains about the following playlist declaration, which we've been successfully using with Liquidsoap 1.4.2: 

```
playlist_sub_burek = playlist(id="playlist_sub_burek",mode="randomize",reload_mode="watch",conservative=true,default_duration=10.,length=20.,"/var/azuracast/stations/lahmacun_radio/playlists/playlist_sub_burek.m3u")
```

Do you have any idea why this is happening?

**Comment by smimram** [comment]
It looks like the standard library is not loaded on your side (and `playlist` is thus not populated).

**Comment by gammaw** [comment]
I knew you would know ;) And what do you think could be the fix? What I did is that I replaced the 1.4.2 liquidsoap file with the newly built one.

**Comment by gammaw** [comment]
Ok, I guess I found what you mean. I'm probably missing these: https://github.com/savonet/liquidsoap/tree/main/libs
I'll try to mount those files too. Thanks for the hint!

**Comment by smimram** [comment]
In case it helps, @toots has added an error when there is no standard library in 7e211ffdb84a5d6b439e7709500a449bfc0c394e.

**Comment by gammaw** [comment]
Thanks, that's a good addition! We were wondering if we could skip compiling the binary ourselves and use a pre-built image instead, run it and copy out the files from it. I did it with the v2.0.0-beta 1 image (from Docker Hub), took the `liquidsoap` binary and the `libs` and mounted them into an `azuracast` image and I got this error

```
liquidsoap: error while loading shared libraries: libgd.so.3: cannot open shared object file: No such file or directory
```

Isn't a liquidsoap binary "stand-alone", which includes all dependencies? Or can it simply be a platform mismatch (OS versions etc.)?

**Comment by toots** [comment]
> Thanks, that's a good addition! We were wondering if we could skip compiling the binary ourselves and use a pre-built image instead, run it and copy out the files from it. I did it with the v2.0.0-beta 1 image (from Docker Hub), took the `liquidsoap` binary and the `libs` and mounted them into an `azuracast` image and I got this error
> 
> ```
> liquidsoap: error while loading shared libraries: libgd.so.3: cannot open shared object file: No such file or directory
> ```
> 
> Isn't a liquidsoap binary "stand-alone", which includes all dependencies? Or can it simply be a platform mismatch (OS versions etc.)?

No, the liquidsoap built from the CI is dynamically loaded. You can try to compile a static binary. I'd recommend `alpine` for that. However, you might not be able to compile all options as some C symbols clash between encoders, typically between ffmpeg and the encoders it already includes, like `libmp3lame`. On the windows build, which is static, we had to only enable the ffmpeg encoder/decoder and disable all the other ones.

**Comment by gammaw** [comment]
Thanks for the insight! Can I use OPAM to build a static binary (https://www.liquidsoap.info/doc-1.4.4/install.html -> Install using OPAM)? And how would one compile a dynamically loaded binary with OPAM?

**Comment by toots** [comment]
I would use the regular `configure` system for this. I think a good starting point would be to be able to build liquidsoap in a alpine OS environment then pass the `-static` linker flag and see what happens..

**Comment by gammaw** [comment]
@toots @smimram feel free to close this issue, I opened another which is more clear and don't mix up issues (operators and installation). Thanks for your help!
